calibre-web-automated-downloader/src/models.rs

```diff
@@ -0,0 +1,217 @@
+            QueueStatus::Available,
+            QueueStatus::Error,
+            QueueStatus::Done,
+        ] {
+            result.insert(status_variant.clone(), HashMap::new());
+        }
+
+        for (book_id, status) in &self.status {
+            if let Some(book_info) = self.book_data.get(book_id) {
+                // Insert a clone so we don't move out of the original book_data
+                result
+                    .get_mut(status)
+                    .unwrap()
+                    .insert(book_id.clone(), book_info.clone());
+            }
+        }
+
+        result
+    }
+
+    /// Refresh the queue by:
+    /// - Checking if "AVAILABLE" books have an .epub file; if not, mark them DONE.
+    /// - Removing stale entries that have exceeded the status_timeout
+    ///   (but only if their status == DONE).
+    pub fn refresh(&mut self) {
+        let now = Instant::now();
+
+        // We'll store which book IDs to update (mark as Done)
+        let mut to_update = Vec::new();
+        // We'll store which book IDs to remove entirely
+        let mut to_remove = Vec::new();
+
+        // First pass: just read immutable data and record changes to apply later
+        for (book_id, status) in &self.status {
+            // If status is AVAILABLE, check if the .epub file does NOT exist => mark as DONE
+            if *status == QueueStatus::Available {
+                let path = CONFIG.ingest_dir.join(format!("{}.epub", book_id));
+                if !path.exists() {
+                    to_update.push(book_id.clone()); // We'll update this after the loop
+                }
+            }
+
+            // Check for stale entries if last update is too old
+            if let Some(ts) = self.status_timestamps.get(book_id) {
+                if now.duration_since(*ts) > self.status_timeout {
+                    // Remove if it's DONE and timed out
+                    if *status == QueueStatus::Done {
+                        to_remove.push(book_id.clone());
+                    }
+                }
+            }
+        }
+
+        // Second pass: apply updates (which requires mutating self).
+        // Now the immutable borrow from the for-loop is no longer in use.
+        for book_id in to_update {
+            self.update_status_internal(&book_id, QueueStatus::Done);
+        }
+
+        // Finally remove stale entries
+        for book_id in to_remove {
+            self.status.remove(&book_id);
+            self.status_timestamps.remove(&book_id);
+            self.book_data.remove(&book_id);
+        }
+    }
+
+    /// Change the status timeout in hours
+    pub fn set_status_timeout(&mut self, hours: u64) {
+        self.status_timeout = Duration::from_secs(hours * 3600);
+    }
+}
+
+// Global instance of BookQueue, protected by a Mutex for thread safety.
+// Access with `BOOK_QUEUE.lock().unwrap()` to get a mutable reference.
+pub static BOOK_QUEUE: Lazy<Mutex<BookQueue>> = Lazy::new(|| {
```

Comment:
If we move the Mutex inside the struct the interfacing with this global variable will be handier.